### PR TITLE
Testing for issued identifiers and poison graphs.

### PR DESCRIPTION
**Pull request by gkellogg**
* [x] Add tests for issued identifiers map `rdfc:RDFC10MapTest`.
* [x] Add tests for poison datasets.

This creates additional test manifests for different types of tests. One for identifiers map looks at JSON results. Forthcoming for poison graphs likely will be an `rdfc:RDFC10NegativeTest` and expect the algorithm to abort due to excessive computation. Just a draft right now.

For `rdfc:RDFC10NegativeEvalTest`, the instructions suggest setting a maximum number of calls to Hash N-Degree Blank Nodes to forty times the total number of blank nodes in the dataset.

### CODE REVIEW

**Comment by iherman** [comment]
(I realize this is still a draft PR): what I see currently is that a map is generated and stored for all tests, which is unnecessary and just confusing. Wouldn't it be better to have them for the mapping tests only? Or do we expect the tests to check for _all_ tests (in which case there is no reason for the separation of the two types of tests...)

**Comment by gkellogg** [comment]
The mechanism for creating these tests favors having two different manifests, we could easily decide on a subset of tests for which to have expected map results, but why not be comprehensive. 

If people would rather see tests merged into a single manifest, we could certainly do that, but numbering and test inputs would be related. 

I do expect the expected error tests to share the same manifest as the canonicalization tests.

**Comment by gkellogg** [comment]
Just to be clear, there are two manifests, on for n-quads canonical output and one for the canonical mapping. Each contains only the tests it uses. It’s driven from the manifest.csv file, and we can change it so that only a subset of the tests look for map outputs, but the input files are shared. We could also put the expected map results in a different sub-directory.

**Comment by iherman** [comment]
Why two manifests? This sounds a bit confusing...

My "mental" model is to have one manifest with different types of test. Depending on the type, the 'result' field refers to an n-quad file or a mapping file, and the test harness could act accordingly. Which also means that the mapping files are present only for those tests that are dedicated to the mapping.

**Comment by gkellogg** [comment]
@iherman the manifests are updated to only have a single top-level manifest (manifest.ttl and manifest.jsonld) which includes all tests, as you suggested. I updated submitted test results because of the new pattern for test entry identifiers.

Most tests have map tests as well. We can, of course, limit this, but I don't see why not include a map test for every input.

**Comment by yamdan** [comment]
Thank you @gkellogg for the PR draft.
I think that the difficulty in introducing the issued identifiers map into test cases lies in the fact that a single test could potentially have more than one correct issued identifiers map.
For instance, test024 could be passed with not just the following expected map:

```json
{ "e0": "c14n0", "e2": "c14n1", "e1": "c14n2"}
```

But also with these two possible expected maps:

```json
{ "e1": "c14n0", "e0": "c14n1", "e2": "c14n2"}
```
```json
{ "e2": "c14n0", "e1": "c14n1", "e0": "c14n2"}
```

(All three of the above maps are consistent with the canonical result: test024-rdfc10.nq)

To validate these additional tests, it seems to me that we have two options:

1. Enumerate all the correct maps. The implementation should be considered to have passed the test if it can output any one of these.
2. Add a method to the specification to define a "canonical" issued identifiers map from multiple correct maps. This could involve, for example, sorting by the input blank node identifiers.

Personally, I believe we should opt for the first approach if the testing mechanism allows. The second option would make the specification dependent on the input blank node identifiers, which does not seem desirable.

**Comment by iherman** [comment]
@gkellogg said:
> Most tests have map tests as well. We can, of course, limit this, but I don't see why not include a map test for every input.

But @yamdan said:
> I think that the difficulty in introducing the issued identifiers map into test cases lies in the fact that a single test could potentially have more than one correct issued identifiers map.
> […]
> To validate these additional tests, it seems to me that we have two options:
> 
> 1. Enumerate all the correct maps. The implementation should be considered to have passed the test if it can output any one of these.
> 2. Add a method to the specification to define a "canonical" issued identifiers map from multiple correct maps. This could involve, for example, sorting by the input blank node identifiers.

I think we overcomplicate things if we follow @gkellogg's comment on including a map test for every output. The type of test cases quoted by @yamdan are, in some respect, "unusual" graphs that do not really occur in practice. By trying to accommodate for those, we do not achieve anything in terms of the goal of the test suite, namely to check the validity of the specification and help developers. I would propose, following the adage "ain't broken don't fix it", to keep the current test suite as is, and _add_ a few test cases whose goals are to test the mappings.

In case we go with a "complete" testing with the mappings provided everywhere, then I'd opt for option (1) of @yamdan, b.t.w.

**Comment by gkellogg** [comment]
Probably the best thing to do, to keep the testing simple, is identify the subset of tests that have unique results and limit the tests to just those input datasets.

**Comment by gkellogg** [comment]
I've removed some unused test files, and eliminated map results which could be ambiguous.

**Comment by gkellogg** [comment]
Added a negative based on a 10-degree clique. My implementation sets a maximum for class to Hash N-Degree Blank Nodes to forty times the number of distinct nodes in the graph (keys in the blank node to quads map). Setting it to twenty caused test048 "evil(3)" to fail. This is rather arbitrary, and we could re-characterized some of the existing stress tests.

File: tests/manifest.csv
Comment (yamdan) [inline tests/manifest.csv:47]:
```suggestion
test058,unnamed graph with blank node objects,,,TRUE,FALSE
```
I think test058 should be removed from rdfc10map tests because it has ambiguous (duplicated) `issued identifiers map`s.

*related manifest files (ttl, jsonld, ...) should also be modified

File: tests/rdfc10/test058-rdfc10map.json
Comment (yamdan) [inline tests/rdfc10/test058-rdfc10map.json:5]: I think this file should be removed because test058 has ambiguous (duplicated) `issued identifiers map`s.

**Review by yamdan — CHANGES_REQUESTED** [review]
Thank you @gkellogg.
I think test058 should also be removed from mapping tests because it has ambiguous `issued identifiers map` as follows:
```json
[ { "e3": "c14n0", "e0": "c14n1", "e1": "c14n2" },
  { "e3": "c14n0", "e1": "c14n1", "e0": "c14n2" } ]
```

**Comment by iherman** [comment]
> 
> The tests all do have unique identifiers, although the naming uses the same base. We did a lot of test duplication in JSON-LD, and it became a pain to maintain. I don't see a real reason to duplicate input files needlessly, but the infrastructure can certainly tolerate tests which are only C14N or Map. At this point, I thought I would go through and designate some of our existing tests which should create unambiguous maps.

I am fine with that structure, +1 to it.

Thanks @yamdan for his extra effort in locating the problematic cases. My implementation does not make it obvious to reveal those, clearly his does...


> 
> > However... those tests seem to test the expected _issued identifiers map_ in the [final step](https://www.w3.org/TR/rdf-canon/#ca.7). Do we want to test the _canonical dataset_ portion of the same step, in particular the blank node identifier map? If yes, how would we do that?
> 
> The **canonicalized dataset** is composed of the original RDF dataset, the **input blank node identifier map** and the **issued identifiers map**. I don't see how we can test the **input blank node identifiers map** as it maps blank nodes (not identifiers) to their identifiers. I'm open to suggestions, but I think what we have speaks to the dual purposes of the algorithm.

This raises a real issue to be whether returning the input bnode identifier map makes sense or not. I will raise a separate issue on this, not to mess up this PR.

**Comment by iherman** [comment]
> Forthcoming for poison graphs likely will be an `rdfc:RDFC10NegativeTest` and expect the algorithm to abort due to excessive computation. Just a draft right now.
>
> For `rdfc:RDFC10NegativeEvalTest`, the instructions suggest setting a maximum number of calls to Hash N-Degree Blank Nodes to forty times the total number of blank nodes in the dataset.

I am not sure how to differentiate between these two tests in practice. Aborting the algorithm would depend on the measure used for excessive computation, and that is also related to the maximum number of calls. None of these are clear...

I could imagine *one* (type of test) which says something like: "you should have a limit which makes this test abort (the limit being implementation dependent)." But this covers both cases, doesn't it?

(see also the open discussion at #120...)

**Comment by iherman** [comment]
For the record: my test harness runs now with the new manifest and the tests (and all are fine, I did not run into the duplicate map issues of @yamdan).

The poison graph is indeed poisonous: I let it run for about 15 minutes, and then I killed it ☹️

**Comment by iherman** [comment]
My proposal at this point would be to merge this PR and handle the poison graphs separately; the exact nature of how to handle them, and how to test them, also depends on #120...

**Comment by yamdan** [comment]
In my implementation, the determination of whether the result of canonicalization can yield multiple `issued identifiers map`s is based on whether there are duplicates in the `hash` in the `hash path list` of Step 5.3 in the [Canonicalization Algorithm](https://w3c.github.io/rdf-canon/spec/#canon-algo-algo).
If there are duplicates, multiple `issued identifiers map`s could potentially exist.

This is because when there are duplicate `hash`es, multiple candidate `result`s that should initially be processed in Step 5.3.1 exist, and the current specification cannot uniquely determine them.
While identifying all these ambiguous points and modifying the specifications would make it possible to obtain a unique `issued identifiers map`, I personally think this would be overkill and exceed the demand.

**Comment by yamdan** [comment]
For instance, consider test058. We find the following two `result`s in the `hash path list`, both having identical `hash` values:
```json
[
  { "hash": "d04073ef64de086a71e72fadb5131832c346ad7b9b5cc71cf8276f2860d48be3",
   "issuer": { "e0": "b0" } },
  { "hash": "d04073ef64de086a71e72fadb5131832c346ad7b9b5cc71cf8276f2860d48be3",
   "issuer": { "e1": "b0" } }
]
```
If Step 5.3.1 processes these results in the presented order, the resulting `issued identifiers map` will be `{ "e3": "c14n0", "e0": "c14n1", "e1": "c14n2" }`.
But, if processed in the reverse order, the map changes to `{ "e3": "c14n0", "e1": "c14n1", "e0": "c14n2" }`.
As far as I can see, the current specification does not provide guidance on determining the processing order when there are duplicate `hash`es at Step 5.3.

File: tests/manifest.csv
Comment (gkellogg) [reply tests/manifest.csv]: Done.

**Comment by gkellogg** [comment]
> @gkellogg what is exactly the meaning of the `poisonfactor` that you introduced in [eb4fff2](https://github.com/w3c/rdf-canon/commit/eb4fff21fe45f26365628ef75e29a2c9a5cd414b)? I _presume_ that it is a multiplicative factor to the number of bnodes, used to determine the maximum number of hash n degree quads calls.

This was requested in the June 27 meeting: [RESOLUTION: remove test 58 (for the map result). Keep 46 but add commentary around it being on the edge but is basically poisonous. Add a new metadata field to flag deliberately evil datasets](https://www.w3.org/2023/06/27-rch-minutes.html#r01). I interpreted the request by determining the multiple of the number of bnodes in dataset needed to do the canonicalization, rather than just say "low", "medium", or "high" degree of toxicity. The [poisonFactor](https://github.com/w3c/rdf-canon/blob/b7f8db16c99daf2126f200c2e0ceda6f8731fcea/tests/vocab.ttl#L61) term is described in the [test vocabulary](https://github.com/w3c/rdf-canon/blob/tests-update/tests/vocab.ttl).

I don't think the actual value matters, but it shows a relative degree of complexity for canonicalizing a dataset. I'm open to other suggestions, and I could interpret the raw value in manifest.csv to use some other value buckets (e.g., "safe", "hard", "toxic").

> **If that is indeed the case:**
> 
> * this should be defined somewhere (I have not found this definition)

See above.

> * it is problematic because it makes the test suite dependent on a particular approach on handling poison graphs… (I happen to use this approach in mine, but that is besides the point...)

It's intended to be informative, and is not needed to provision a test runner. The instructions in the README (and generated index.html) should be sufficient.

>   (I think we agreed that implementations _MUST_ have some defense measure against poison graphs, but it did not say what that is. See all the discussion in [What a reasonable way of defending against poison graphs? #120](https://github.com/w3c/rdf-canon/issues/120))

From the README:

> For a negative evaluation test, the test passes if the implementation generates an error due to excessive calls to [Hash N-Degree Quads](https://www.w3.org/TR/rdf-canon/#hash-nd-quads-algorithm). The maximum number of calls that should be allowed for running these tests is forty times the total number of blank nodes in the input dataset.

> I believe that the test should be formulated by saying that it must have a way to report an error, but we should not (cannot) go beyond that in the test suite imho.

That description says it must generate an error. It doesn't say how it should do this, or any message that should accompany it. It's up to each implementation to decide how to do this and report.

**Comment by iherman** [comment]
> > @gkellogg what is exactly the meaning of the `poisonfactor` that you introduced in [eb4fff2](https://github.com/w3c/rdf-canon/commit/eb4fff21fe45f26365628ef75e29a2c9a5cd414b)? I _presume_ that it is a multiplicative factor to the number of bnodes, used to determine the maximum number of hash n degree quads calls.
> 
> This was requested in the June 27 meeting: [RESOLUTION: remove test 58 (for the map result). Keep 46 but add commentary around it being on the edge but is basically poisonous. Add a new metadata field to flag deliberately evil datasets](https://www.w3.org/2023/06/27-rch-minutes.html#r01). I interpreted the request by determining the multiple of the number of bnodes in dataset needed to do the canonicalization, rather than just say "low", "medium", or "high" degree of toxicity. The [poisonFactor](https://github.com/w3c/rdf-canon/blob/b7f8db16c99daf2126f200c2e0ceda6f8731fcea/tests/vocab.ttl#L61) term is described in the [test vocabulary](https://github.com/w3c/rdf-canon/blob/tests-update/tests/vocab.ttl).

Ah, o.k., I was not present on the call and did not read the minutes. My apologies.

I think the numerical value is really misleading because silly people may want to read into it more than it is intended. I think a simple enumeration of fixed terms ("none", "medium", and "high" seems o.k. to me) is perfectly enough and fine.

Thanks

**Comment by TallTed** [comment]
> I think the numerical value is really misleading because silly people may want to read into it more than it is intended. I think a simple enumeration of fixed terms ("none", "medium", and "high" seems o.k. to me) is perfectly enough and fine.

I agree with this.

**Comment by gkellogg** [comment]
There’s the [Poisoning Severity Score](https://cdn.who.int/media/docs/default-source/chemical-safety/intox/pss95699a36-61ab-4be6-848f-c1d894d21fbd.pdf?sfvrsn=5750967e_10&download=true) from the CDC. NONE, MINOR, MODERATE, SEVERE, DEATH.

**Comment by gkellogg** [comment]
Changed "poisonFactor" to "computationalComplexity", "low", "medium", or "high".

* "low" – The multiplier on blank nodes is 0.
* "medium" – The multiplier on blank nodes is from 1 to 10.
* "high" – The multiplier on blank nodes is greater than 10.

Controlled by [this code](https://github.com/w3c/rdf-canon/blob/19ae4ea691419622371650d4712aab7be5c9c90c/tests/mk_manifest.rb#L53-L60): 

    def computational_complexity
      case self.complexity.to_i
      when 0 then 'low'
      when 1..10 then 'medium'
      else 'high'
      end
    end

**Comment by gkellogg** [comment]
I believe I've addressed all comments. Some approval reviews would be good, or point out what I've missed.

**Review by yamdan — COMMENTED** [review]
Thank you @gkellogg for all your hard work.
I have confirmed that my implementation can pass all of the latest test cases, including map tests and negative tests.
The following suggestions are mainly related to the test vocabulary. I apologize for the delay.

File: tests/README.md
Comment (iherman) [inline tests/README.md:28]: This reads like a constraint on the algorithm specification. It should rather be formulated in terms of the test structure, something like "all blank nodes appearing in the test appear in...".

File: tests/template.haml
Comment (iherman) [inline tests/template.haml:80]: (This seems to be a copy of the README file, see my comments there...)

**Review by iherman — APPROVED** [review]

File: tests/vocab.ttl
Comment (iherman) [reply tests/vocab.ttl:62]: I am not sure... we refer to poison graphs in the main text, so keeping the same term here sounds o.k. to me. We can make it more verbose (it is a label, after all), but saying: "Poison dataset (measuring the reaction on computational complexity)."

File: tests/vocab.jsonld
Comment (iherman) [reply tests/vocab.jsonld:46]: Actually... even adding a `"@type":"rdfs:Class"` sounds unnecessary. The `subClassOf` suffices.

File: tests/vocab.jsonld
Comment (iherman) [reply tests/vocab.jsonld:53]: See my comment above

File: tests/vocab.jsonld
Comment (iherman) [reply tests/vocab.jsonld:60]: See my comment above...

**Comment by iherman** [comment]
> Thank you @gkellogg for all your hard work.

Indeed, thanks @gkellogg 

> I have confirmed that my implementation can pass all of the latest test cases, including map tests and negative tests.

Same here.

Once this is merged, I will have to issue a new update of my library to npm, and I will then generate a new, 'official' earl file.

File: tests/vocab.ttl
Comment (gkellogg) [reply tests/vocab.ttl:62]: The degree of complexity can make a graph poison, which is noted in the `rdfs:comment`.

File: tests/vocab.jsonld
Comment (gkellogg) [reply tests/vocab.jsonld:46]: Not sure why I made that change, I'm going to keep it as `rdfs:Class`, even though it can be inferred.

File: tests/mk_manifest.rb
Comment (gkellogg) [reply tests/mk_manifest.rb:20]: `template.haml` takes them from `mk_manifest.rb`.

File: tests/README.md
Comment (gkellogg) [reply tests/README.md:28]: I'll save this for subsequent clarification and more specific wording. It is true that the values of one map are the keys of the other, and they should match exactly; this isn't a constraint on the algorithm.

File: tests/vocab.ttl
Comment (yamdan) [inline tests/vocab.ttl:44]:
```suggestion
    Canonicalization performed using the RDFC-1.0 algorithm.
    Results are compared as text
    where the result of running the test is serialized to canonical N-Quads, sorted in code point order.
```
I think the following phrase located from line 36 to 37 would be better relocated here because it no longer applies to the test as a whole, but is now only applicable to the EvalTest.

> Results are compared as text where the result of running the test is serialized to canonical N-Quads, lexicographically-sorted.

(However, it appears that I cannot make a suggestion for lines 36-37...)

**Review by yamdan — COMMENTED** [review]
Thank you so much @gkellogg , I just added a small suggestion for `vocab.ttl`.
Additionally, while I could not directly suggest in this PR review, it seems that lines 30 and 31 of `tests/vocab_template.haml` should be revised as follows:

(Before)
```
        [Turtle](vocab.ttl) and [JSON-LD](vocab.jsonld) versions of the vocabular are also available.
        The vocabulary is published by the [W3C Credentials Community Group](https://www.w3.org/community/credentials/).
```

(After)
```
        [Turtle](vocab.ttl) and [JSON-LD](vocab.jsonld) versions of the vocabulary are also available.
        The vocabulary is published by the [W3C RDF Dataset Canonicalization Working Group](https://www.w3.org/groups/wg/rch).
```
- vocabular -> vocabulary
- CCG -> RCH WG

**Review by yamdan — APPROVED** [review]
Thanks!

**Review by iherman — APPROVED** [review]
Just to make it clear: all my pending comments are minor and editorial ones on the descriptions. This can be finalized later; it is important to merge this PR asap to complete the test suite. So +1 to merge right now.

**Comment by iherman** [comment]
Yay!

**Pull request merged**
